runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/TheatreOfBloodPlugin.java

```diff
@@ -0,0 +1,472 @@
+
+	@Inject
+	private TheatreOfBloodConfig config;
+
+	private final List<Attempt> attempts = new ArrayList<>();
+	private Attempt current;
+	// Room stats
+	private double dealt = 0;
+	private double taken = 0;
+
+	private int state = 0;
+	private int region = 0;
+	private double hpExp = 0;
+	private Actor oldTarget;
+	private boolean isSpectator = false;
+
+	@Provides
+	TheatreOfBloodConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(TheatreOfBloodConfig.class);
+	}
+
+	@Override
+	protected void shutDown()
+	{
+		attempts.clear();
+		state = -1;
+	}
+
+	@Subscribe
+	protected void onChatMessage(ChatMessage m)
+	{
+		if (!isSpectator && state > 1 && m.getType() == ChatMessageType.SERVER)
+		{
+			String message = Text.removeTags(m.getMessage());
+			Matcher match = DEATH_TEXT.matcher(message);
+			if (match.matches())
+			{
+				// Died
+				current.addDeath();
+			}
+		}
+	}
+
+	@Subscribe
+	protected void onVarbitChanged(VarbitChanged e)
+	{
+		int old = state;
+		state = client.getVar(Varbits.THEATRE_OF_BLOOD);
+		if (old != state)
+		{
+			stateChanged(old);
+		}
+	}
+
+	@Subscribe
+	protected void onHitsplatApplied(HitsplatApplied e)
+	{
+		// Hitsplat on us while inside Theatre of Blood means we took damage
+		if (state == 2 && e.getActor().equals(client.getLocalPlayer()))
+		{
+			current.addDamageTaken(e.getHitsplat().getAmount());
+			taken += e.getHitsplat().getAmount();
+		}
+	}
+
+	@Subscribe
+	protected void onExperienceChanged(ExperienceChanged c)
+	{
+		if (c.getSkill().equals(Skill.HITPOINTS) && state >= 2)
+		{
+			// Calculate XP difference for HP skill and calculate damage based off it
+			double oldExp = hpExp;
+			hpExp = client.getSkillExperience(Skill.HITPOINTS);
+			double diff = hpExp - oldExp;
+			if (diff < 1)
+			{
+				return;
+			}
+			double damageDealt = calculateDamageDealt(diff);
+
+			// Add damage dealt to the current logs
+			log.debug("Damage Dealt: {} | Exact: {}", Math.round(damageDealt), damageDealt);
+			current.addDamageDealt(damageDealt);
+			dealt += damageDealt;
+		}
+	}
+
+	@Subscribe
+	public void onGameStateChanged(GameStateChanged event)
+	{
+		if (!isSpectator && state >= 1 && event.getGameState() == GameState.LOGGED_IN)
+		{
+			// LOGGED_IN is triggered between region changes
+			int oldRegion = region;
+			region = WorldPoint.fromLocalInstance(client, client.getLocalPlayer().getLocalLocation()).getRegionID();
+
+			if (region != oldRegion)
+			{
+				handleRegionChange();
+			}
+		}
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick tick)
+	{
+		if (!isSpectator && state == 2)
+		{
+			oldTarget = client.getLocalPlayer().getInteracting();
+		}
+	}
+
+	/**
+	 * Handles changes to the Theatre of Blood Varbit
+	 * @param old previous Theatre of Blood varbit value
+	 */
+	private void stateChanged(int old)
+	{
+		// TODO: Figure out a way to determine if they are logging back into a raid
+		// 0=default | 1=party | 2=inside/spectator | 3=dead-spectator
```

Comment:
Just to re-iterate the above - Once you're at the stage you're describing what your ints mean with comments, use an enum ;)